src/tests/ftest/cart/util/daos_sys_logscan.py

```diff
@@ -0,0 +1,561 @@
+    def _is_leader(self, puuid, rank, pid):
+        if puuid not in self._pools:
+            return False
+        if self._cur_ldr_rank[puuid] == rank and self._cur_ldr_pid[puuid] == pid:
+            return True
+        return False
+
+    def _get_ps_leader_components(self, match):
+        # puuid, term
+        # see re_step_up and re_step_down
+        return match.group(1), int(match.group(2))
+
+    def _match_ps_step_up(self, fname, line, pid, rank):
+        msg, host, datetime = self._get_line_components(line)
+        match = self.re_step_up.match(msg)
+        if match:
+            puuid, term = self._get_ps_leader_components(match)
+            if puuid not in self._pools:
+                self._pools[puuid] = {}
+            self._cur_ldr_rank[puuid] = rank
+            self._cur_ldr_pid[puuid] = pid
+            self._cur_term[puuid] = term
+            old_term = term - 1
+            # if term already exists, error?
+            if term in self._pools[puuid]:
+                self._warn(f"pool {puuid} term {term} already seen!", fname, line)
+            # carry over most recent map version into the new term, avoid later KeyError
+            if old_term in self._pools:
+                if self._pools and self._pools[puuid][old_term]["maps"] != {}:
+                    last_mapver = max(self._pools[puuid][old_term]["maps"].keys())
+                    pmap_versions = self._pools[puuid][old_term]["maps"][last_mapver]
+                    pmap_versions["carryover"] = True
+            else:
+                pmap_versions = {}
+            self._pools[puuid][term] = {
+                "rank": rank,
+                "begin_time": datetime,
+                "end_time": "",
+                "host": host,
+                "pid": pid,
+                "logfile": fname,
+                "maps": pmap_versions
+            }
+            if self._debug:
+                print(f"{datetime} FOUND pool {puuid} BEGIN\tterm {term} pmap_versions empty: "
+                      f"{str(pmap_versions == {})} rank {rank}\t{host}\tPID {pid}\t{fname}")
+            return True
+        return False
+
+    def _match_ps_step_down(self, fname, line, pid, rank):
+        msg, host, datetime = self._get_line_components(line)
+        match = self.re_step_down.match(msg)
+        if match:
+            puuid, term = self._get_ps_leader_components(match)
+            if term != self._cur_term[puuid]:
+                self._warn(f"step_down term={term} != cur_term={self._cur_term}", fname, line)
+            self._cur_ldr_rank[puuid] = -1
+            self._cur_ldr_pid[puuid] = -1
+            self._cur_term[puuid] = -1
+            self._pools[puuid][term]["end_time"] = datetime
+            if self._debug:
+                print(f"{datetime} FOUND pool {puuid} END\tterm {term} rank {rank}\t{host}\t"
+                      f"PID {pid}\t{fname}")
+            return True
+        return False
```

Comment:
Recommend reversing the `if` - handling exit condition first
```suggestion
        match = self.re_step_down.match(msg)
        if not match:
            return False
        puuid, term = self._get_ps_leader_components(match)
        if term != self._cur_term[puuid]:
            self._warn(f"step_down term={term} != cur_term={self._cur_term}", fname, line)
        self._cur_ldr_rank[puuid] = -1
        self._cur_ldr_pid[puuid] = -1
        self._cur_term[puuid] = -1
        self._pools[puuid][term]["end_time"] = datetime
        if self._debug:
            print(f"{datetime} FOUND pool {puuid} END\tterm {term} rank {rank}\t{host}\t"
                  f"PID {pid}\t{fname}")
        return True
```

Reply:
done in  latest commit